src/Logging.js

```diff
@@ -0,0 +1,380 @@
+        // keep track of the new level
+        _debugLevelFunctionSet = parseDebugLevelInput(newDebugLevel);
+
+        // the highest (least restrictive debug level) always wins
+        _curDebugLevel = Math.max(Math.max(_debugLevelQueryString, _debugLevelFunctionSet), _curDebugLevel);
+    } catch (e) {}
+}
+
+function okToLogMessage(level) {
+    return level <= _curDebugLevel; // getReqestedMaxDebugLevel();
+}
+
+// DEPRECATED
+function tryLogMessageLegacy(level, message, source) {
+    try {
+        var messageToLog = '[APN-' + level + '-' + new Date().toISOString() + '] ';
+        if (source !== null && source && source.length > 0) {
+            messageToLog += source + '>';
+        }
+        messageToLog += message;
+
+        if (okToLogMessage(level)) {
+            console.log(messageToLog);
+        }
+    } catch (ex) {
+        if (okToLogMessage(level)) {
+            console.log(ex);
+        }
+    }
+}
+
+module.exports = {
+
+    /**
+     * Call the appropriate trace method at the given level
+     * @param (string) debugLevel = Level to debug at
+     */
+    traceAtLevel: function() {
+        try {
+            if (arguments.length > 0) {
+                var targetTraceLevel = arguments[0];
+                var argsWithoutTraceLevel = Array.prototype.slice.call(arguments, 1);
+                traceMessageAtLevel.call(this, targetTraceLevel, argsWithoutTraceLevel);
+            }
+        } catch (e) {}
+    },
+
+    /**
+     * If the logging level for type "always" or higher is set, output message to browser's console log
+     * With the [APN] prefix and styled with the browser's console "log" level trace style
+     * @see https://developer.mozilla.org/en-US/docs/Web/API/console#Outputting_text_to_the_console
+     * for more details about how to use console.log style logging.
+     */
+    always: function() {
+        try {
+            traceMessageAtLevel.call(this, TRACE_LEVEL_ALWAYS, Array.prototype.slice.call(arguments));
+        } catch (e) {}
+    },
+
+    /**
+     * If the logging level for type "error" or higher is set, output message to browser's console log
+     * With the [APN] prefix and styled with the browser's console "error" level trace style
+     * @see https://developer.mozilla.org/en-US/docs/Web/API/console#Outputting_text_to_the_console
+     * for more details about how to use console.error style logging.
+     */
+    error: function() {
+        try {
+            traceMessageAtLevel.call(this, TRACE_LEVEL_ERROR, Array.prototype.slice.call(arguments));
+        } catch (e) {}
+    },
+
+    /**
+     * If the logging level for type "log" or higher is set, output message to browser's console log
+     * With the [APN] prefix and styled with the browser's console "log" level trace style
+     * @see https://developer.mozilla.org/en-US/docs/Web/API/console#Outputting_text_to_the_console
+     * for more details about how to use console.log style logging.
+     */
+    log: function() {
```

Comment:
Can this just simplify to `traceMessageAtLevel.bind(this, TRACE_LEVEL_LOG)`? Same with the other log-level functions.

You'd have to move the `try/catch` inside `traceMessageAtLevel`, but.... I don't see any calls to it which aren't wrapping it in a `try/catch` anyway, so... that should be safe.

Reply:
The traceMessageAtLevel already has try/catch inside. I think I can easilly remove try/catch around any invocation of this function. Is this OK?

Reply:
oh, duh... dunno how I missed that. Yeah, that works. There was just a lot of repeated boilerplate here.

I think `.bind` should be more efficient, but you're right, it does subtly change the API from expanded args to a single array of args. `traceMessageAtLevel is private, so it can be fixed later easily enough.